state-chain/amm-math/src/lib.rs

```diff
@@ -348,59 +247,267 @@ pub fn tick_at_sqrt_price(sqrt_price: SqrtPriceQ64F96) -> Tick {
+	let c: U512 = c.into();
+	(U256::full_mul(a, b) / c).try_into().ok()
+}
+
+pub fn mul_div_ceil<C: Into<U512>>(a: U256, b: U256, c: C) -> U256 {
+	mul_div(a, b, c).1
+}
+
+pub fn mul_div<C: Into<U512>>(a: U256, b: U256, c: C) -> (U256, U256) {
+	let c: U512 = c.into();
+
+	let (d, m) = U512::div_mod(U256::full_mul(a, b), c);
+
+	(
+		d.try_into().unwrap(),
+		if m > U512::from(0) {
+			// cannot overflow as for m > 0, c must be > 1, and as (a*b) < U512::MAX, therefore
+			// a*b/c < U512::MAX
+			d + 1
+		} else {
+			d
+		}
+		.try_into()
+		.unwrap(),
 	)
-	.0 >> 128u8)
-		.as_u128() as Tick; // Add Checks
+}
 
-	if tick_low == tick_high {
-		tick_low
-	} else if sqrt_price_at_tick(tick_high) <= sqrt_price {
-		tick_high
-	} else {
-		tick_low
+#[derive(
+	PartialEq,
+	Eq,
+	PartialOrd,
+	Ord,
+	Debug,
+	Clone,
+	Copy,
+	Default,
+	Encode,
+	Decode,
+	Serialize,
+	Deserialize,
+	TypeInfo,
+	MaxEncodedLen,
+)]
+/// This is the ratio of equivalently valued amounts of asset One and asset Zero.
+///
+/// The price is always measured in amount of asset One per unit of asset Zero. Therefore as asset
+/// zero becomes more valuable relative to asset one the prices literal value goes up, and vice
+/// versa. This ratio is represented as a fixed point number with `PRICE_FRACTIONAL_BITS` fractional
+/// bits.
+pub struct Price(U256);
+
+impl From<SqrtPriceQ64F96> for Price {
+	fn from(sqrt_price: SqrtPriceQ64F96) -> Self {
+		assert!(sqrt_price.is_valid());
+
+		// Note the value here cannot ever be zero as MIN_SQRT_PRICE has its 33th bit set, so
+		// sqrt_price will always include a bit pass the 64th bit that is set, so when we shift
+		// down below that set bit will not be removed.
+		Price(mul_div_floor(
+			sqrt_price.0,
+			sqrt_price.0,
+			U256::one() << (2 * SqrtPriceQ64F96::FRACTIONAL_BITS - Price::FRACTIONAL_BITS),
+		))
 	}
 }
 
+impl Price {
+	pub const FRACTIONAL_BITS: u32 = 128;
+
+	pub fn from_raw<T: Into<U256>>(value: T) -> Self {
+		Price(value.into())
+	}
+
+	pub fn as_raw(self) -> U256 {
+		self.0
+	}
+
+	pub const fn zero() -> Self {
+		Price(U256::zero())
+	}
+
+	pub fn is_zero(&self) -> bool {
+		self.0.is_zero()
+	}
+
+	/// Converts a `tick` to a `price`. Will return `None` for ticks outside MIN_TICK..=MAX_TICK
+	///
+	/// This function never panics.
+	pub fn from_tick(tick: Tick) -> Option<Self> {
+		if is_tick_valid(tick) {
+			Some(SqrtPriceQ64F96::from_tick(tick).into())
+		} else {
+			None
+		}
+	}
+
+	pub fn from_amounts_bounded(quote: Amount, base: Amount) -> Self {
+		SqrtPriceQ64F96::from_amounts_bounded(quote, base).into()
+	}
+
+	/// Given prices of asset 1 and asset 2 (in terms of the same asset)
+	/// compute the price of asset 1 (self) in terms of asset 2 (given)
+	pub fn relative_to(self, price: Price) -> Self {
+		Price(mul_div_floor(self.0, U256::one() << Self::FRACTIONAL_BITS, price.0))
+	}
+
+	pub fn output_amount_floor(self, input: Amount) -> Amount {
```

Comment:
Methods like this are where using trait abstractions makes more sense, so that the caller doesn't need to be concerned with the inner U256 type at all:

```
pub fn output_amount_floor<I: Into<U256>, O: TryFrom<U256>>(self, input: I) -> O {
```